### DISCUSSION TITLE
How does a PCI device perform virtual to physical address translation before executing DMA in an x86 full system?

### DISCUSSION THREAD

**Discussion by ferry-hhh** [gem5/gem5#1301] [Q&A] [answered]
Generally speaking, PCI devices use IOMMU (for x86) or SMMU (for ARM) for virtual to physical address translation. In the current version of gem5, I noticed that there is indeed an SMMU for the ARM architecture, but I couldn't find a similar component for x86. I have examined the DMA process for typical PCI devices such as copy_engine and ide under x86, but I don't understand how they perform the virtual to physical address translation.
I would like to ask for your help in understanding how PCI devices under x86 perform virtual to physical address translation. Or, can I use the SMMU in an x86 system? I am mainly focusing on full system simulations.
Thank you in advance for your assistance and guidance.

**Comment by ritaru** [comment]
Take a look at src/dev/pci/host.hh:289 ([Link](https://github.com/gem5/gem5/blob/bb418d41eb6d87c0a0591869097005c16420c6aa/src/dev/pci/host.hh#L289)).

**Reply by ferry-hhh** [reply]
Sorry for the late reply. First of all, thank you very much for your answer. I have seen the `pioAddr()` and `dmaAddr()` functions you mentioned before, but I understand that the simple addition in the above functions should not be able to complete the complex task of virtual-to-physical address translation. Your answer made me revisit this part of the code.

I see that typical PCI devices such as copy engines and IDE devices run `pciToDma(address)` when executing [dmaAction()](https://github.com/gem5/gem5/blob/bb418d41eb6d87c0a0591869097005c16420c6aa/src/dev/pci/copy_engine.cc#L463), as shown below:
```c++
    cePort.dmaAction(MemCmd::ReadReq, ce->pciToDma(address),
                     sizeof(DmaDesc), &fetchCompleteEvent,
                     (uint8_t*)curDmaDesc, latBeforeBegin);
```
Logically, `pciToDma` is performing the virtual-to-physical address translation. Further tracing the implementation of this function, it ultimately executes [dmaAddr()](https://github.com/gem5/gem5/blob/bb418d41eb6d87c0a0591869097005c16420c6aa/src/dev/pci/host.hh#L297), as follows:
``` c++
    Addr dmaAddr(const PciBusAddr &bus_addr, Addr pci_addr) const override {
        return pciDmaBase + pci_addr;
    }
```

I subsequently ran a full system printout of `address` and `pciToDma(address)` and found that the two are the same, both being physical addresses, so `pciToDma` should not be the place where the actual virtual-to-physical address translation is completed. So, I wanted to trace where the source of this physical address is. Taking the IDE device as an example:
When the IDE device executes `dmaRead()`, it also uses [pciToDma()](https://github.com/gem5/gem5/blob/bb418d41eb6d87c0a0591869097005c16420c6aa/src/dev/storage/ide_disk.cc#L434), for example:
```c++
        ctrl->dmaRead(pciToDma(dmaReadCG->addr()), dmaReadCG->size(),
                &dmaReadWaitEvent, dataBuffer + dmaReadCG->complete());
```

The `dmaReadCG->addr()` is obtained by some conversion of the base address saved in the [PrdTableEntry](https://github.com/gem5/gem5/blob/bb418d41eb6d87c0a0591869097005c16420c6aa/src/dev/storage/ide_disk.hh#L79) class, which should be equivalent to a descriptor in a network card, recording the source address, target address, and the amount of data for the DMA task. The base address of the `PrdTableEntry` is saved in the BMIRegs of the IDE controller. The host driver may write the base address of the `PrdTableEntry` to the BMIRegs via PIO, and the IDE Controller reads back the `PrdTableEntry` and performs the actual DMA task based on the information in the `PrdTableEntry`.

So I understand that the physical address used by the IDE device for DMA might be pre-converted by the driver in the kernel.

I am not sure if my understanding is correct. If there are any mistakes, please correct me.

**Comment by abmerop** [comment, marked as answer]
As far as I know, there are no x86 devices implemented which need virtual-to-physical translation besides the amdgpu model, so IOMMU has similarly not been implemented.

**Reply by ferry-hhh** [reply]
Thank you very much for your answer!
